integration/integration_test.go

```diff
@@ -67,7 +89,598 @@ func getIntegrationTest(numNodes, numVotes int) func(*testing.T) {
+		require.NoError(t, err)
+
+		// ##### DECRYPT BALLOTS #####
+		t.Logf("decrypting")
+
+		form, err = getForm(formFac, formID, nodes[0].GetOrdering())
+		t.Logf("PubsharesUnit: %v", form.PubsharesUnits)
+		require.NoError(t, err)
+		err = decryptBallots(m, actor, form)
+		require.NoError(t, err)
+
+		err = waitForStatus(types.ResultAvailable, formFac, formID, nodes,
+			numNodes, 1500*time.Millisecond*time.Duration(numVotes))
+		require.NoError(t, err)
+
+		t.Logf("get vote proof")
+		form, err = getForm(formFac, formID, nodes[0].GetOrdering())
+		require.NoError(t, err)
+
+		fmt.Println("Title of the form : " + form.Configuration.MainTitle)
+		fmt.Println("ID of the form : " + string(form.FormID))
+		fmt.Println("Status of the form : " + strconv.Itoa(int(form.Status)))
+		fmt.Println("Number of decrypted ballots : " + strconv.Itoa(len(form.DecryptedBallots)))
+
+		//require.Len(t, form.DecryptedBallots, numVotes-numBadVotes)
+		//should contains numBadVotes empty ballots
+		count := 0
+		for _, ballot := range form.DecryptedBallots {
+			if ballotIsNull(ballot) {
+				count++
+			}
+			//fmt.Println(fmt.Sprintf("%#v", ballot))
+		}
+		fmt.Println(form.DecryptedBallots)
+
+		require.Equal(t, numBadVotes, count)
+
+		fmt.Println("closing nodes")
+
+		err = closeNodes(nodes)
+		require.NoError(t, err)
+
+		fmt.Println("test done")
+	}
+}
+
+func getIntegrationTestRevote(numNodes, numVotes int) func(*testing.T) {
+	return func(t *testing.T) {
+		t.Parallel()
+
+		adminID := "first admin"
+
+		// ##### SETUP ENV #####
+		// make tests reproducible
+		rand.Seed(1)
+
+		delaPkg.Logger = delaPkg.Logger.Level(zerolog.WarnLevel)
+
+		dirPath, err := os.MkdirTemp(os.TempDir(), "d-voting-three-votes")
+		require.NoError(t, err)
+
+		defer os.RemoveAll(dirPath)
+
+		t.Logf("using temp dir %s", dirPath)
+
+		// ##### CREATE NODES #####
+		nodes := setupDVotingNodes(t, numNodes, dirPath, nil)
+
+		signer := createDVotingAccess(t, nodes, dirPath)
+
+		m := newTxManager(signer, nodes[0], time.Second*time.Duration(numNodes/2+1), numNodes*4)
+
+		err = grantAccess(m, signer)
+		require.NoError(t, err)
+
+		for _, n := range nodes {
+			err = grantAccess(m, n.GetShuffleSigner())
+			require.NoError(t, err)
+		}
+
+		// ##### CREATE FORM #####
+		formID, err := createForm(m, "Three votes form", adminID)
+		require.NoError(t, err)
+
+		time.Sleep(time.Second * 1)
+
+		// ##### SETUP DKG #####
+		actor, err := initDkg(nodes, formID, m.m)
+		require.NoError(t, err)
+
+		// ##### OPEN FORM #####
+		err = openForm(m, formID)
+		require.NoError(t, err)
+
+		formFac := types.NewFormFactory(types.CiphervoteFactory{}, nodes[0].GetRosterFac())
+
+		t.Logf("start casting votes")
+		form, err := getForm(formFac, formID, nodes[0].GetOrdering())
+		require.NoError(t, err)
+
+		_, err = castVotesRandomly(m, actor, form, numVotes)
+		require.NoError(t, err)
+
+		castedVotes, err := castVotesRandomly(m, actor, form, numVotes)
+		require.NoError(t, err)
```

Comment:
the number of times votes are re-cast could be an argument or a random number. With the current implementation it doesn't make a difference, but it's always better to define stronger invariants in case we were to change that implementation later.